lgc/builder/BuilderImpl.cpp

```diff
@@ -503,36 +521,130 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
+
+  // Insert new code just before nonUniformInst.
+  SetInsertPoint(nonUniformInst);
+
+  // The first begin contains a null token for the previous token argument
+  waterfallBegin = ConstantInt::get(getInt32Ty(), 0);
+  for (auto nonUniformVal : nonUniformIndices)
+    // Start the waterfall loop using the waterfall index.
+    waterfallBegin = CreateIntrinsic(Intrinsic::amdgcn_waterfall_begin, nonUniformVal->getType(),
+                                     {waterfallBegin, nonUniformVal}, nullptr, instName);
+
+  // Scalarize each non-uniform operand of the instruction.
+  for (unsigned operandIdx : operandIdxs) {
+    Value *desc = nonUniformInst->getOperand(operandIdx);
+    auto descTy = desc->getType();
+#if LLVM_MAIN_REVISION && LLVM_MAIN_REVISION < 463892
+    // Old version of the code
+#else
+    // When the non-uniform use is in a VGPR, we can save a v_mov by not inserting the amdgcn_waterfall_readfirstlane
+    if (!useVgprForOperands)
+#endif
+    desc = CreateIntrinsic(Intrinsic::amdgcn_waterfall_readfirstlane, {descTy, descTy}, {waterfallBegin, desc}, nullptr,
+                           instName);
+    if (nonUniformInst->getType()->isVoidTy()) {
+      // The buffer/image operation we are waterfalling is a store with no return value. Use
+      // llvm.amdgcn.waterfall.last.use on the descriptor.
+#if LLVM_MAIN_REVISION && LLVM_MAIN_REVISION < 463892
+      // Old version of the code
+      desc = CreateIntrinsic(Intrinsic::amdgcn_waterfall_last_use, descTy, {waterfallBegin, desc}, nullptr, instName);
+#else
+      desc = CreateIntrinsic(useVgprForOperands ? Intrinsic::amdgcn_waterfall_last_use_vgpr
+                                                : Intrinsic::amdgcn_waterfall_last_use,
+                             descTy, {waterfallBegin, desc}, nullptr, instName);
+#endif
+    }
+    // Replace the descriptor operand in the buffer/image operation.
+    nonUniformInst->setOperand(operandIdx, desc);
   }
+  return waterfallBegin;
+}
+
+Value *BuilderImpl::emitWaterfallBeginForScalarizedLoops(Instruction *nonUniformInst, Instruction *firstIndexInst,
+                                                         ArrayRef<unsigned> operandIdxs,
+                                                         ArrayRef<Value *> nonUniformIndices,
+                                                         DenseMap<Value *, SmallVector<Value *, 2>> loadChain,
+                                                         bool useVgprForOperands, const Twine &instName) {
+
+  if (loadChain.empty())
+    return emitWaterfallBegin(nonUniformInst, operandIdxs, nonUniformIndices, useVgprForOperands, instName);
+
+  // Check if there is a load chain for all the operands.
+  for (unsigned operandIdx : operandIdxs) {
+    Value *desc = nonUniformInst->getOperand(operandIdx);
+    if (!loadChain.contains(desc))
+      return emitWaterfallBegin(nonUniformInst, operandIdxs, nonUniformIndices, useVgprForOperands, instName);
+  }
+
+  assert(firstIndexInst);
```

Comment:
Assert for both `nonUniformInst` and `firstIndexInst`?

Reply:
I added it. I am not sure if it is needed.